lib/requestExecutor.js

```diff
@@ -0,0 +1,360 @@
+    } catch (error) {
+        logger.error(`Redis connection error: ${error}`);
+    }
+}
+
+const cortexId = config.get('cortexId');
+const connection = client && new Bottleneck.IORedisConnection({ client: client });
+
+let modelEndpoints = {};
+
+const createLimiter = (endpoint, name, index) => {
+    const rps = endpoint.requestsPerSecond ?? 100;
+    let limiterOptions = {
+        minTime: 1000 / rps,
+        maxConcurrent: rps,
+        reservoir: rps,      // Number of tokens available initially
+        reservoirRefreshAmount: rps,     // Number of tokens added per interval
+        reservoirRefreshInterval: 1000, // Interval in milliseconds
+    };
+
+    // If Redis connection exists, add id and connection to enable clustering
+    if (connection) {
+        limiterOptions.id = `${cortexId}-${name}-${index}-limiter`; // Unique id for each limiter
+        limiterOptions.connection = connection;  // Shared Redis connection
+        }
+
+    endpoint.limiter = new Bottleneck(limiterOptions);
+
+    endpoint.limiter.on('error', (err) => {
+        logger.error(`Limiter error for ${cortexId}-${name}-${index}: ${err}`);
+        endpoint.limiter.disconnect();
+        createLimiter(endpoint, name, index);
+        logger.info(`New limiter created for ${cortexId}-${name}-${index}`)
+    });
+
+    endpoint.limiter.on('failed', (error, info) => {
+        if (error.name === 'CanceledError') {
+            logger.debug(`Request cancelled for ${cortexId}-${name}-${index}: Id: ${info.options.id || 'none'}`);
+        } else {
+            logger.error(`Request failed for ${cortexId}-${name}-${index}: Id: ${info.options.id || 'none'}: ${error}`);
+        }
+    });
+
+    endpoint.limiter.on('debug', (message) => {
+        if (!message.includes('heartbeat.lua')) {
+            logger.debug(`Limiter ${cortexId}-${name}-${index}: ${message}`);
+        }
+    });
+}
+
+const buildModelEndpoints = (config) => {
+    modelEndpoints = JSON.parse(JSON.stringify(config.get('models')));
+    logger.info(`Building ${connection ? 'Redis clustered' : 'local'} model rate limiters for ${cortexId}...`);
+    for (const [name, model] of Object.entries(modelEndpoints)) {
+        model.endpoints.forEach((endpoint, index) => {
+            createLimiter(endpoint, name, index)
+            endpoint.monitor = new RequestMonitor();
+        });
+    }
+}
+
+let currentIndex = 0; // for round-robin selection
+
+const selectEndpoint = (model) => {
+    if (!model || !Array.isArray(model.endpoints) || model.endpoints.length === 0) {
+        return null;
+    } else {
+        logger.debug(`Selecting endpoint for model ${model.name}...`);
+        if (model.endpoints.length === 1) {
+            logger.debug(`Only one endpoint for model ${model.name}. No selection required.`);
+            return model.endpoints[0];
+        }
+
+        let healthyEndpoints = model.endpoints.filter(endpoint => endpoint.monitor.healthy);
```

Comment:
i think this is great for the first version. i'd still vote for some kind of pq approach then we could possibly get rid of several ifs and pick in o1, but just a thought